distributed/security.py

```diff
@@ -1,18 +1,107 @@
+
 __all__ = ("Security",)
 
 
+if sys.version_info >= (3, 10) or _OPENSSL_VERSION_INFO >= (1, 1, 0, 7):
+
+    def _create_default_context(
+        purpose: ssl.Purpose,
+        *,
+        cafile: StrOrBytesPath | None = None,
+        capath: StrOrBytesPath | None = None,
+        cadata: str | bytes | None = None,
+        minimum_version: ssl.TLSVersion | None,
+        maximum_version: ssl.TLSVersion | None,
+    ) -> ssl.SSLContext:
+        # The OP_NO_SSL* and OP_NO_TLS* become deprecated in favor of
+        # 'SSLContext.minimum_version' from Python 3.7 onwards, however
+        # this attribute is not available unless the ssl module is compiled
+        # with OpenSSL 1.1.0g or newer.
+        # https://docs.python.org/3.10/library/ssl.html#ssl.SSLContext.minimum_version
+        # https://docs.python.org/3.7/library/ssl.html#ssl.SSLContext.minimum_version
+        context = ssl.create_default_context(
+            purpose, cafile=cafile, capath=capath, cadata=cadata
+        )
+        if minimum_version is not None:
+            context.minimum_version = minimum_version
+        if maximum_version is not None:
+            context.maximum_version = maximum_version
+        return context
+
+else:
+    _min_version = {
+        None: 0,
+        ssl.TLSVersion.MINIMUM_SUPPORTED: ~ssl.OP_NO_SSLv2 | ~ssl.OP_NO_SSLv3,
+        ssl.TLSVersion.SSLv3: ssl.OP_NO_SSLv2 | ~ssl.OP_NO_SSLv3,
+        ssl.TLSVersion.TLSv1: ssl.OP_NO_SSLv2 | ssl.OP_NO_SSLv3,
+        ssl.TLSVersion.TLSv1_1: ssl.OP_NO_SSLv2 | ssl.OP_NO_SSLv3 | ssl.OP_NO_TLSv1,
+        ssl.TLSVersion.TLSv1_2: ssl.OP_NO_SSLv2
+        | ssl.OP_NO_SSLv3
+        | ssl.OP_NO_TLSv1
+        | ssl.OP_NO_TLSv1_1,
+        ssl.TLSVersion.TLSv1_3: ssl.OP_NO_SSLv2
+        | ssl.OP_NO_SSLv3
+        | ssl.OP_NO_TLSv1
+        | ssl.OP_NO_TLSv1_1
+        | ssl.OP_NO_TLSv1_2,
+    }
+    _max_version = {
+        None: 0,
+        ssl.TLSVersion.TLSv1_3: 0,
+        ssl.TLSVersion.TLSv1_2: ssl.OP_NO_TLSv1_3,
+        ssl.TLSVersion.TLSv1_1: ssl.OP_NO_TLSv1_2 | ssl.OP_NO_TLSv1_3,
+        ssl.TLSVersion.TLSv1: ssl.OP_NO_TLSv1_1 | ssl.OP_NO_TLSv1_2 | ssl.OP_NO_TLSv1_3,
+        ssl.TLSVersion.SSLv3: ssl.OP_NO_SSLv3
+        | ssl.OP_NO_TLSv1_1
+        | ssl.OP_NO_TLSv1_2
+        | ssl.OP_NO_TLSv1_3,
+    }
+
+    def _create_default_context(
+        purpose: ssl.Purpose,
+        *,
+        cafile: StrOrBytesPath | None = None,
+        capath: StrOrBytesPath | None = None,
+        cadata: str | bytes | None = None,
+        minimum_version: ssl.TLSVersion | None,
+        maximum_version: ssl.TLSVersion | None,
+    ) -> ssl.SSLContext:
+        context = ssl.create_default_context(
+            purpose, cafile=cafile, capath=capath, cadata=cadata
+        )
+        try:
+            context.options |= (
+                _min_version[minimum_version] | _max_version[maximum_version]
+            )
+        except KeyError:
+            raise RuntimeError("invalid TLSVersion selected")
```

Comment:
setting maximum_version=MAXIMUM_SUPPORTED or maximum_version=MINIMUM_SUPPORTED or minimum_version=MAXIMUM_SUPPORTED isn't possible because we can't introspect the versions until 1.1.0.g anyway